README
Test/Pod version 1.53
=====================

This library's module, Test::Pod, provides an interface for validating POD in
module files.

INSTALLATION

To install this module, type the following:

    perl Makefile.PL
    make
    make test
    make install

Dependencies
------------

Test::Pod requires the following modules:

* Pod::Simple 3.05
* Test::More 0.62
* Test::Builder::Tester 1.02
* File::Spec

Copyright and Licence
---------------------

Copyright 2006-2010, Andy Lester; 2010-2015 David E. Wheeler. Some Rights
Reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.


Changes
Change log for Test::Pod

    * *.psgi files are now recognized as pod-containing files (thanks, Tom
      Hukins!)
    * explicit perl 5.8 dependency removed

1.51  2015-07-04 20:36:53Z
    * convert tooling from Module::Build to ExtUtils::MakeMaker (resolves
      circular dependency with MB on perl 5.21+)

1.50  2015-05-28T21:42:48Z
    * Restored support for ignoring directories listed in
      `%Test::Pod::ignore_dirs`, inadvertently dropped by the switch to
      File::Find in v1.46. Thanks to Julien ÉLIE for the report and diagnosis!

1.49  2015-05-12T17:30:58Z
    * Changed license in the README to "Same as Perl" to match the license
      listed in `META.yml`, and in the POD as of v1.42.
    * Tightened up the heuristic for matching Perl batch files to require
      that the file end in `.bat`, as described in the documentation.
      Patch from Luca Ferrari.
    * Simplified some of the matching code. Patch from Luca Ferrari.
    * Transferred ownership of the repository to the perl-pod organization on
      GitHub.
    * Removed the Pod testing test from the distribution. Seems silly for a
      module designed for testing Pod, right? But it's considered best
      practice to do Pod testing in development only, not in a release.

1.48 - 2013-05-06T04:47:00Z
    * Fixed test failure with versions of Pod::Simple lower than 3.24. Thanks
      to Martin Holste and Tatsuhiko Miyagawa for the reports.

1.47 - 2013-05-05T03:14:25Z
    * Fixed failing test on Pod::Simple 3.27 and higher.

1.46 - 2013-02-16T19:46:17Z
    * Fixed a crash when `all_pod_files_ok()` finds no files to test. Thanks
      to H.Merijn Brand.
    * Switched from custom file searching code to File::Find. Suggested by
      H.Merijn Brand.
    * Fixed failing test on Pod::Simple 3.24.

1.45 - 2011-03-09T20:00:59
    * Removed Module::Build from `build_requires`, since it does nothing
      out-of-the ordinary, and this keeps it out of the `Makefile.PL`'s
      `PM_PREREQ`. Per gripe from Peter Scott.

1.44 - 2010-04-26T17:33:02
    * Restored the `Makefile.PL` by using Module::Build::Compat's
      "traditional" configuration.

1.43 - 2010-04-21T02:33:59
    * Reduced prereq for Test::More to 0.62, prereq for Pod::Simple to 3.05,
      and added File::Spec to the list of prereqs. The prerequisites were
      needlessly high, and by reducing them Test::Pod can rely on the core
      versions of those modules in on Perl 5.8.8. Thanks to Chris Wittmer for
      the patch.

1.42 - 2010-03-10T19:41:20
    * Fixed invalid Pod link. Thanks to Markus Sonderegger.
    * Changed license in the POD to "Same as Perl" to match the license listed in
      `META.yml`.
    * Changed `Build.PL` to require Module::Build 0.30. It already does in
      `configure_requires`, but also doing so on the `use` line helps provide
      a more useful error message in older toolchain tools that don't use
      `configure_requires`.
    * Updated `MANIFEST` so that the `README` is actually included in the
      distribution.
    * Now recognizes `.bat` files as Perl files when they contain
      "--*-Perl-*--" on the first line. Thanks to Olivier 'dolmen' Mengué for
      the patch (RT #46973).
    * `all_pod_files_ok()` now operates on directories as well as files.
      Thanks to Adriano Ferreira for the patch (RT #33025).
    * Files with no POD now have " (no pod)" added to test name. Thanks to
      Adriano Ferreira for the patch (RT #34955).
    * Updated documentation to note that `.PL` files are considered Perl
      files. Thanks to Adriano Ferreira for the spot (RT #34955).
    * `all_pod_files_ok()` now lets `pod_file_ok()` set the default test name,
      rather than doing so itself. Thanks to Adriano Ferreira for the spot (RT
      #34955).

1.41 - 2010-01-14T20:09:25
    * Maintenance transfered to David Wheeler.
    * Test::Pod no longer complains about the construct L<text|url>, as it is
      no longer illegal (as of Perl 5.11.3).
    * Switched to Module::Build.

1.40 - Sun Jul 12 23:32:11 CDT 2009
    [THINGS THAT MAY BREAK YOUR CODE]
    Test::Pod now requires Perl 5.8.0.

    [ENHANCEMENTS]
    Test::Pod now complains about the illegal construct L<text|url>.
    Thanks to Paul Miller.

    The list of directories to exclude is now much longer, and is
    available in %Test::Pod::ignore_dirs.  This list is right now:

        '.bzr' => 'Bazaar',
        '.git' => 'Git',
        '.hg'  => 'Mercurial',
        '.pc'  => 'quilt',
        '.svn' => 'Subversion',
        CVS    => 'CVS',
        RCS    => 'RCS',
        SCCS   => 'SCCS',
        _darcs => 'darcs',
        _sgbak => 'Vault/Fortress',


1.26 - Wed Jul 19 09:54:48 CDT 2006

    No new functionality in this version.

    [FIXES]
    * Fixed a Win32 build bug.
    * Fixed a bug in the SYNOPSIS.

1.24 - Wed Feb  1 15:18:06 PST 2006
    [THINGS THAT MAY BREAK YOUR CODE]
    * The long-deprecated pod_ok() is now gone, along with all its
      constants.

    [ENHANCEMENTS]
    * Now runs properly under VMS.  Thanks to Peter Edwards.

1.22 - Sun Oct 23 23:45:51 CDT 2005
    [ENHANCEMENTS]
    * Added check for lines with only whitespace before or after a
      directive.  Older POD formatters can get confused by such
      lines since they goof up paragraph mode. (RT #6467)

    [FIXES]
    * Bumped up requirements to Test::Simple 0.62.

1.20 - Wed Jun 23 00:28:35 CDT 2004
    [ENHANCEMENTS]
    * Looks in blib/ if there is one, otherwise looks in lib/
    * Thanks to David Wheeler for nudging.

1.18 - Fri May 28 23:48:28 CDT 2004
    [FIXES]
    * Now exports all_pod_files() as advertised.
    * Also includes *.t files as Pod.

1.16 - Fri Apr 30 17:42:50 CDT 2004
    [FIXES]
    * Fixed a warning under 5.8.3.  Also explicitly closes dir handles
      when searching for POD.

1.14 - Wed Apr 28 23:38:29 CDT 2004
    [ENHANCEMENTS]
    * Now runs taint-safe.  No longer uses File::Find.  I'm trying to
      get all my modules to run under -T correctly, so this is big.


lib/Test/Pod.pm
package Test::Pod;

use strict;
use warnings;

=head1 NAME

Test::Pod - Check for POD errors in files

=head1 VERSION

Version 1.53

=cut

our $VERSION = '1.53';

=head1 SYNOPSIS

C<Test::Pod> lets you check the validity of a POD file, and report
its results in standard C<Test::Simple> fashion.

    use Test::Pod tests => $num_tests;
    pod_file_ok( $file, "Valid POD file" );

Module authors can include the following in a F<t/pod.t> file and
have C<Test::Pod> automatically find and check all POD files in a
module distribution:

    use Test::More;
    eval "use Test::Pod 1.00";
    plan skip_all => "Test::Pod 1.00 required for testing POD" if $@;
    all_pod_files_ok();

You can also specify a list of files to check, using the
C<all_pod_files()> function supplied:

    use strict;
    use Test::More;
    eval "use Test::Pod 1.00";
    plan skip_all => "Test::Pod 1.00 required for testing POD" if $@;
    my @poddirs = qw( blib script );
    all_pod_files_ok( all_pod_files( @poddirs ) );

Or even (if you're running under L<Apache::Test>):

    use strict;
    use Test::More;
    eval "use Test::Pod 1.00";
    plan skip_all => "Test::Pod 1.00 required for testing POD" if $@;

    my @poddirs = qw( blib script );
    use File::Spec::Functions qw( catdir updir );
    all_pod_files_ok(
        all_pod_files( map { catdir updir, $_ } @poddirs )
    );

=head1 DESCRIPTION

Check POD files for errors or warnings in a test file, using
C<Pod::Simple> to do the heavy lifting.

=cut

use Test::Builder;
use Pod::Simple;

our %ignore_dirs = (
    '.bzr' => 'Bazaar',
    '.git' => 'Git',
    '.hg'  => 'Mercurial',
    '.pc'  => 'quilt',
    '.svn' => 'Subversion',
    CVS    => 'CVS',
    RCS    => 'RCS',
    SCCS   => 'SCCS',
    _darcs => 'darcs',
    _sgbak => 'Vault/Fortress',
);

my $Test = Test::Builder->new;

sub import {
    my $self = shift;
    my $caller = caller;

    for my $func ( qw( pod_file_ok all_pod_files all_pod_files_ok ) ) {
        no strict 'refs';
        *{$caller."::".$func} = \&$func;
    }

    $Test->exported_to($caller);
    $Test->plan(@_);
}

sub _additional_test_pod_specific_checks {
    my ($ok, $errata, $file) = @_;

    return $ok;
}

=head1 FUNCTIONS

=head2 pod_file_ok( FILENAME[, TESTNAME ] )

C<pod_file_ok()> will okay the test if the POD parses correctly.  Certain
conditions are not reported yet, such as a file with no pod in it at all.

When it fails, C<pod_file_ok()> will show any pod checking errors as
diagnostics.

The optional second argument TESTNAME is the name of the test.  If it
is omitted, C<pod_file_ok()> chooses a default test name "POD test
for FILENAME".

=cut

sub pod_file_ok {
    my $file = shift;
    my $name = @_ ? shift : "POD test for $file";

    if ( !-f $file ) {
        $Test->ok( 0, $name );
        $Test->diag( "$file does not exist" );
        return;
    }

    my $checker = Pod::Simple->new;

    $checker->output_string( \my $trash ); # Ignore any output
    $checker->parse_file( $file );

    my $ok = !$checker->any_errata_seen;
       $ok = _additional_test_pod_specific_checks( $ok, ($checker->{errata}||={}), $file );

    $name .= ' (no pod)' if !$checker->content_seen;
    $Test->ok( $ok, $name );
    if ( !$ok ) {
        my $lines = $checker->{errata};
        for my $line ( sort { $a<=>$b } keys %$lines ) {
            my $errors = $lines->{$line};
            $Test->diag( "$file ($line): $_" ) for @$errors;
        }
    }

    return $ok;
} # pod_file_ok

=head2 all_pod_files_ok( [@entries] )

Checks all the files under C<@entries> for valid POD. It runs
L<all_pod_files()> on directories and assumes everything else to be a file to
be tested. It calls the C<plan()> function for you (one test for each file),
so you can't have already called C<plan>.

If C<@entries> is empty or not passed, the function finds all POD files in
files in the F<blib> directory if it exists, or the F<lib> directory if not.
A POD file matches the conditions specified below in L</all_pod_files>.

If you're testing a module, just make a F<t/pod.t>:

    use Test::More;
    eval "use Test::Pod 1.00";
    plan skip_all => "Test::Pod 1.00 required for testing POD" if $@;
    all_pod_files_ok();

Returns true if all pod files are ok, or false if any fail.

=cut

sub all_pod_files_ok {
    my @args = @_ ? @_ : _starting_points();
    my @files = map { -d $_ ? all_pod_files($_) : $_ } @args;

    unless (@files) {
        $Test->skip_all( "No files found in (@args)\n" );
        return 1;
    }

    $Test->plan( tests => scalar @files );

    my $ok = 1;
    foreach my $file ( @files ) {
        pod_file_ok( $file ) or undef $ok;
    }
    return $ok;
}

=head2 all_pod_files( [@dirs] )
X<all_pod_files>

Returns a list of all the POD files in I<@dirs> and in directories below. If
no directories are passed, it defaults to F<blib> if F<blib> exists, or else
F<lib> if not. Skips any files in F<CVS>, F<.svn>, F<.git> and similar
directories. See C<%Test::Pod::ignore_dirs> for a list of them.

A POD file is:

=over 4

=item * Any file that ends in F<.pl>, F<.PL>, F<.pm>, F<.pod>, F<.psgi> or F<.t>.

=item * Any file that has a first line with a shebang and "perl" on it.

=item * Any file that ends in F<.bat> and has a first line with "--*-Perl-*--" on it.

=back

The order of the files returned is machine-dependent.  If you want them
sorted, you'll have to sort them yourself.

=cut

sub all_pod_files {
    my @pod;
    require File::Find;
    File::Find::find({
        preprocess => sub { grep {
            !exists $ignore_dirs{$_}
            || !-d File::Spec->catfile($File::Find::dir, $_)
        } @_ },
        wanted   => sub { -f $_ && _is_perl($_) && push @pod, $File::Find::name },
        no_chdir => 1,
    }, @_ ? @_ : _starting_points());
    return @pod;
}

sub _starting_points {
    return 'blib' if -e 'blib';
    return 'lib';
}

sub _is_perl {
    my $file = shift;

    # accept as a Perl file everything that ends with a well known Perl suffix ...
    return 1 if $file =~ /[.](?:PL|p(?:[lm]|od|sgi)|t)$/;

    open my $fh, '<', $file or return;
    my $first = <$fh>;
    close $fh;
    return unless $first;

    # ... or that has a she-bang as first line ...
    return 1 if $first =~ /^#!.*perl/;

    # ... or that is a .bat ad has a Perl comment line first
    return 1 if $file =~ /[.]bat$/i && $first =~ /--[*]-Perl-[*]--/;

    return;
}

=head1 SUPPORT

This module is managed in an open L<GitHub
repository|http://github.com/perl-pod/test-pod/>. Feel free to fork and
contribute, or to clone L<git://github.com/perl-pod/test-pod.git> and send
patches!

Found a bug? Please L<post|http://github.com/perl-pod/test-pod/issues> or
L<email|mailto:bug-test-pod@rt.cpan.org> a report!

=head1 AUTHORS

=over

=item David E. Wheeler <david@justatheory.com>

Current maintainer.

=item Andy Lester C<< <andy at petdance.com> >>

Maintainer emeritus.

=item brian d foy

Original author.

=back

=head1 ACKNOWLEDGEMENTS

Thanks brian d foy for the original code, and to these folks for contributions:

=over

=item * Andy Lester

=item * David E. Wheeler

=item * Paul Miller

=item * Peter Edwards

=item * Luca Ferrari

=back

=head1 COPYRIGHT AND LICENSE

Copyright 2006-2010, Andy Lester; 2010-2015 David E. Wheeler. Some Rights
Reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut

1;


Makefile.PL
use strict;
use warnings;
require 5.008;
use ExtUtils::MakeMaker;

my $developer = -f '.gitignore';
if ($developer) {
    ExtUtils::MakeMaker->VERSION('6.98');
    require ExtUtils::Manifest;
    ExtUtils::Manifest->VERSION('1.70');
}

my %WriteMakefileArgs = (
    NAME                => 'Test::Pod',
    VERSION_FROM        => 'lib/Test/Pod.pm',
    ABSTRACT_FROM       => 'lib/Test/Pod.pm',
    AUTHOR              => 'David E. Wheeler <david@justatheory.com>',
    LICENSE             => 'perl_5',

    META_MERGE => {
        'meta-spec' => { version => 2 },
        dynamic_config => 0,
        resources => {
            homepage => 'http://search.cpan.org/dist/Test-Pod/',
            license => ['http://dev.perl.org/licenses/'],
            repository => {
                url => 'https://github.com/perl-pod/test-pod.git',
                web => 'https://github.com/perl-pod/test-pod',
                type => 'git',
            },
            bugtracker => {
                web => 'http://github.com/perl-pod/test-pod/issues/',
            },
        },
        x_MailingList => 'http://lists.perl.org/list/pod-people-qa.html',
        x_IRC => 'irc://irc.perl.org/#perl-qa', # maybe we should make a pod channel?
        x_contributors => [ # manually added, from git shortlog -e -s -n
            'David E. Wheeler <david@justatheory.com>',
            'Andy Lester <andy@petdance.com>',
            'Paul Miller <jettero@cpan.org>',
            'Luca Ferrari <fluca1978@gmail.com>',
            'Karen Etheridge <ether@cpan.org>',
            'ChrisWi <chris@computersalat.de>',
            'David Steinbrunner <dsteinbrunner@pobox.com>',
        ],
    },

    META_ADD => {
        'meta-spec' => { version => 2 },
        prereqs => {
            configure => {
                requires => {
                    'ExtUtils::MakeMaker' => '0',
                },
            },
            runtime => {
                requires => {
                    'perl'                  => '5.008',
                    'File::Find'            => '0',
                    'Pod::Simple'           => '3.05',
                    'Test::Builder::Tester' => '1.02',
                    'Test::More'            => '0.62',
                },
            },
            test => {
                requires => {
                    'File::Spec'    => '0',
                    'Pod::Simple'   => '3.05',
                    'Test::More'    => '0.62',
                },
            },
        },
    },
);

my $eumm_version  = eval $ExtUtils::MakeMaker::VERSION;

for (qw(configure build test runtime)) {
    my $key = $_ eq 'runtime' ? 'PREREQ_PM' : uc $_.'_REQUIRES';
    next unless exists $WriteMakefileArgs{META_ADD}{prereqs}{$_}
             or exists $WriteMakefileArgs{$key};
    my $r = $WriteMakefileArgs{$key} = {
        %{$WriteMakefileArgs{META_ADD}{prereqs}{$_}{requires} || {}},
        %{delete $WriteMakefileArgs{$key} || {}},
    };
    defined $r->{$_} or delete $r->{$_} for keys %$r;
}

# dynamic prereqs get added here.

$WriteMakefileArgs{MIN_PERL_VERSION} = delete $WriteMakefileArgs{PREREQ_PM}{perl} || 0;

die 'attention developer: you need to do a sane meta merge here!'
    if keys %{$WriteMakefileArgs{BUILD_REQUIRES}};

$WriteMakefileArgs{BUILD_REQUIRES} = {
    %{$WriteMakefileArgs{BUILD_REQUIRES} || {}},
    %{delete $WriteMakefileArgs{TEST_REQUIRES}}
} if $eumm_version < 6.63_03;

$WriteMakefileArgs{PREREQ_PM} = {
    %{$WriteMakefileArgs{PREREQ_PM}},
    %{delete $WriteMakefileArgs{BUILD_REQUIRES}}
} if $eumm_version < 6.55_01;

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
    if $eumm_version < 6.51_03;

delete $WriteMakefileArgs{MIN_PERL_VERSION}
    if $eumm_version < 6.48;

delete @WriteMakefileArgs{qw(META_ADD META_MERGE)}
    if $eumm_version < 6.46;

delete $WriteMakefileArgs{LICENSE}
    if $eumm_version < 6.31;

WriteMakefile(%WriteMakefileArgs);


Build.PL
use strict;
use warnings;
use Module::Build '0.30';

Module::Build->new(
    module_name        => 'Test::Pod',
    license            => 'perl',
    dist_author => 'David E. Wheeler <david@justatheory.com>',
    configure_requires => { 'Module::Build' => '0.30', },
    create_makefile_pl => 'traditional',
    build_requires     => {
        'File::Spec'    => 0,
        'Pod::Simple'   => '3.05',
        'Test::More'    => '0.62',
    },
    requires => {
        'File::Find'            => 0,
        'Pod::Simple'           => '3.05',
        'Test::Builder::Tester' => '1.02',
        'Test::More'            => '0.62',
    },
    meta_merge => {
        resources => {
            homepage   => 'http://search.cpan.org/dist/Test-Pod/',
            bugtracker => 'http://github.com/perl-pod/test-pod/issues/',
            repository => 'http://github.com/perl-pod/test-pod',
        }
    },
)->create_build_script;
